### DISCUSSION TITLE
2. 네트워크 - 전송 계층

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#156] [질의 응답]
전송 계층 질의응답

**Comment by aeroej** [comment]
## Q. TCP와 UDP의 주요 차이점은 무엇인가요?

TCP는 데이터 전송의 신뢰성을 보장합니다. 오류 복구와 흐름 제어 기능이 있습니다. 
UDP는 데이터 전송의 신뢰성은 보장하지 않지만, TCP보다 데이터 전송 속도가 빠릅니다.

**Reply by sikutisa** [reply]
## Q. IP 프로토콜은 신뢰성을 보장하지 않습니다. 그런데 TCP는 어떻게 신뢰성을 보장하나요?
* 세그먼트 개념을 사용합니다.
* 누적 ACK을 사용합니다.
* time-out 개념을 사용합니다.
* 재전송(retransmission)을 사용합니다.

**Comment by watty-exclamationMark** [comment]
## Q. 전송 계층의 역할은 무엇인가요?

데이터 분할, 에러 관리, 흐름 제어, 연결 설정과 관련된 역할을 합니다.

**Reply by sikutisa** [reply]
## Q. 전송계층과 네트워크 계층의 차이는 뭔가요?
* 전송계층에서는 프로세스 사이에서 논리적인 통신이 이뤄지고, 네트워크 계층은 호스트 사이에서 논리적인 통신이 이뤄집니다.
* (각주) 논리적인 통신이라는 점을 주의하자

**Comment by watty-exclamationMark** [comment]
## Q. TCP 3-way Handshake란 무엇인가요?
TCP 연결 설정 절차로, 송신자와 수신자의 연결을 확인하는 방법입니다.
3단계의 절차로 이루어집니다.
1. 송신 측에서 연결을 요청합니다 (SYN)
2. 목적지에서 연결을 수락하고 응답합니다. (SYN-ACK)
3. 송신 측에서 연결을 수락하고 응답합니다. (ACK)

**Comment by sikutisa** [comment]
## Q. ACK(acknowledgement)란 무엇인가요?
* 수신자가 송신자에게, 패킷을 이상없이 정상적으로 받았다고 알려주기 위한 패킷입니다.
* 반대 개념으로 NACK이 있습니다.
* (각주) link-layer ACK, end-to-end ACK에 대해 찾아보면 좋다

**Reply by sikutisa** [reply]
## Q. Timeout 란 무엇인가요?
* 메시지를 보낸 이후, 일정 시간동안 ACK이나 NACK을 받지 못하면 timeout이 발생하게 됩니다.
* (각주) rdt 2.0, rdt 2.1, rdt 2.2 rdt 3.0에 대해 찾아보면 좋다. 특히 rdt 3.0

**Reply by sikutisa** [reply]
## Q. TCP Sequence Number에 대한 개념과, ACK과의 관계에 대해 설명하세요.
* 네트워크상에서 데이터가 전송될 때, TCP는 전체 데이터를 MSS 단위로 쪼갠 세그먼트들을 구성해 전송하게 됩니다.
* 각 세그먼트 데이터가 네트워크에서 온전히 전송된다는 보장이 없습니다. 특정 데이터는 소실될 수도 있습니다.
* 따라서 TCP는 몇번째 세그먼트까지는 전송에 성공했고, 어떤 세그먼트는 재전송할 필요가 있는지 판단할 필요가 있는데 Sequence Number는 이때 사용됩니다.
* ACK에는 수신자가 다음에 수신할 것으로 예상하는 세그먼트의 번호가 담겨서 전송됩니다. 이를 통해 TCP는 어떤 세그먼트가 유실됐는지를 판단할 수 있습니다
* (각주) Go-Back-N, Selective Repeat에 대해 찾아보자.
